src/art/art.c

```diff
@@ -0,0 +1,1622 @@
+            return art_node16_erase((art_node16_t *)node, key_chunk);
+        case ART_NODE48_TYPE:
+            return art_node48_erase((art_node48_t *)node, key_chunk);
+        case ART_NODE256_TYPE:
+            return art_node256_erase((art_node256_t *)node, key_chunk);
+        default:
+            assert(false);
+            return NULL;
+    }
+}
+
+// Inserts the leaf with the given key chunk in the inner node, returns a
+// pointer to the (possibly expanded) node.
+static art_node_t *art_node_insert_leaf(art_inner_node_t *node,
+                                        art_key_chunk_t key_chunk,
+                                        art_leaf_t *leaf) {
+    art_node_t *child = (art_node_t *)(SET_LEAF(leaf));
+    switch (art_get_type(node)) {
+        case ART_NODE4_TYPE:
+            return art_node4_insert((art_node4_t *)node, child, key_chunk);
+        case ART_NODE16_TYPE:
+            return art_node16_insert((art_node16_t *)node, child, key_chunk);
+        case ART_NODE48_TYPE:
+            return art_node48_insert((art_node48_t *)node, child, key_chunk);
+        case ART_NODE256_TYPE:
+            return art_node256_insert((art_node256_t *)node, child, key_chunk);
+        default:
+            assert(false);
+            return NULL;
+    }
+}
+
+// Frees the node and its children. Leaves are freed by the user.
+static void art_free_node(art_node_t *node) {
+    if (art_is_leaf(node)) {
+        // We leave it up to the user to free leaves.
+        return;
+    }
+    switch (art_get_type((art_inner_node_t *)node)) {
+        case ART_NODE4_TYPE:
+            art_free_node4((art_node4_t *)node);
+            break;
+        case ART_NODE16_TYPE:
+            art_free_node16((art_node16_t *)node);
+            break;
+        case ART_NODE48_TYPE:
+            art_free_node48((art_node48_t *)node);
+            break;
+        case ART_NODE256_TYPE:
+            art_free_node256((art_node256_t *)node);
+            break;
+        default:
+            assert(false);
+    }
+}
+
+// Returns the next child in key order, or NULL if called on a leaf.
+// Provided index may be in the range [-1, 255].
+static art_indexed_child_t art_node_next_child(const art_node_t *node,
+                                               int index) {
+    if (art_is_leaf(node)) {
+        art_indexed_child_t indexed_child;
+        indexed_child.child = NULL;
+        return indexed_child;
+    }
+    switch (art_get_type((art_inner_node_t *)node)) {
+        case ART_NODE4_TYPE:
+            return art_node4_next_child((art_node4_t *)node, index);
+        case ART_NODE16_TYPE:
+            return art_node16_next_child((art_node16_t *)node, index);
+        case ART_NODE48_TYPE:
+            return art_node48_next_child((art_node48_t *)node, index);
+        case ART_NODE256_TYPE:
+            return art_node256_next_child((art_node256_t *)node, index);
+        default:
+            assert(false);
+            return (art_indexed_child_t){};
```

Comment:
Is this type of return statement portable?

Reply:
Changed to `return (art_indexed_child_t){0};`